FPGA: Add sample for streaming data interfaces with pipes

williamwang256 opened the pull request:
# Adding a New Sample(s)

## Description

A new sample demonstrating how to use pipes to implement streaming data interfaces on IP components.

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [ ] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

williamwang256 committed:
add code

williamwang256 committed:
rename files to work with regtest

williamwang256 committed:
rename to streaming data interfaces

williamwang256 committed:
fix typos and rename file in cmake

williamwang256 commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+### Reading the Reports
+
+After compiling in the report flow, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pa ...
```
The report says the interface has a width of 32 bits (= `sizeof(StreamingBeatT)`), but the documentation says _"Using the StreamingBeat struct with the uses_packets template parameter set to true adds two additional 1-bit signals to the Avalon interface, start_of_packet (sop), and end_of_packet (eop)."_ Would that not mean the interface should be 16 + 2 = 18 bits wide? or does it do some packing?

With an 8-bit-wide data type the report shows a width of 24 bits.

yuguen replied:
Something to clarify with eng

whitepau replied:
that sounds like a report defect to me.
the report should show the interface width as 16 bits wide (since you declared `StreamingBeatT` to have `ac_int<16, false>` as the underlying type).

If an 8-bit-wide data type results in the report showing a 24-bit interface, it sounds like those 1-bit signals are being parsed inside the compiler as 8-bit signals, and those are trimmed down after the report is generated.

williamwang256 replied:
Okay. I will open a case on this

williamwang256 commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,117 @@
+  // Pipe properties
+  using PipePropertiesT = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::ready_latency<0>,
+    sycl::ext::intel::experimental::bits_per_symbol<16>,
```
Using a 16-bit-wide type while the issue with 8-bit-wide types is being fixed. should I set bits_per_symbol to be 16 to match (so I don't need to add the empty signal)? or should I keep it at 8 and add the empty signal

whitepau replied:
you don't have to add the empty signal, since in this case there will never be an empty symbol.

for tidiness you should probably set bits_per_symbol to 16 though

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,259 @@
+target_link_libraries(${REPORT_TARGET} ${MODIFIED_COMMON_LINK_FLAGS_REPORT})
+target_link_libraries(${REPORT_TARGET} ${REPORT_LINK_FLAGS})
+set_target_properties(${REPORT_TARGET} PROPERTIES OUTPUT_NAME ${REPORT_OUTPUT_NAME})
+
```
Why no FPGA target?

williamwang256 replied:
forgot to add it back... fixed now:)

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,259 @@
+###############################################################################
+### Generate Report
+###############################################################################
+add_executable(${REPORT_TARGET} ${SOURCE_FILES})
```
```suggestion
add_executable(${REPORT_TARGET} EXCLUDE_FROM_ALL ${SOURCE_FILES})
```

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,259 @@
+###############################################################################
+### FPGA Simulator
+###############################################################################
+add_executable(${SIMULATOR_TARGET} ${SOURCE_FILES})
```
```suggestion
add_executable(${SIMULATOR_TARGET} EXCLUDE_FROM_ALL ${SOURCE_FILES})
```

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+Each individual pipe is a function scope class declaration of the templated `pipe` class.
+- The first template parameter is a user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the  RTL.
+- The second template parameter defines the datatype of elements carried by the interface.
+> :warning: There is currently a known issue with using struct types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (e.g., `unsigned char`). For the time being, please use a wider datatype where applicable (e.g., 16-bits).
```
I'm guessing that there is a case for this. Please add a note to the case that when addressed, this warning should be removed.

whitepau replied:
case is here: https://hsdes.intel.com/appstore/article/#/14019166650

I added a reminder about this code sample.

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+#### Table 1. Properties used to Configure a Pipe to Implement a Streaming Data Interface
+
+| Property | Valid Values | Default Value |
```
Please align the table so that the column aligns even in textual form

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
````diff
@@ -0,0 +1,272 @@
+      ```
+   4. Compile for FPGA hardware (longer compile time, targets an FPGA device).
+      ```
+      make fpga
````
The corresponding cmake target is missing

williamwang256 replied:
fixed

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/sample.json:
```diff
@@ -0,0 +1,71 @@
+          "cd build",
+          "cmake ..",
+          "make fpga_emu",
+          "./threshold_packets.fpga_emu"
```
threshold_packets? why not streaming_interfaces?

williamwang256 replied:
Was the old name of the sample. Fixed now

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,117 @@
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/pipes_ext.hpp>
+#include <sycl/sycl.hpp>
+
```
It's missing the exception handler header, and the it's associated property to be added to the queue

yuguen reviewed (changes requested):
I think we need to discuss where we want to put this sample as we have more and more IPA interfaces samples. Maybe it would be better to create an "interfaces" folder or something.
We'll chat about that in the next code samples users group

williamwang256 committed:
Apply suggestions from code review

Co-authored-by: yuguen-intel <yohann.uguen@intel.com>

williamwang256 committed:
addressing comments from Yohann's review

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,292 @@
+# Display the compile instructions in the emulation flow
+getCompileCommands("${COMMON_COMPILE_FLAGS}" "${EMULATOR_COMPILE_FLAGS}" "${COMMON_LINK_FLAGS}" "${EMULATOR_LINK_FLAGS}" "${EMULATOR_TARGET}" "${EMULATOR_OUTPUT_NAME}${EXECUTABLE_EXTENSION}")
+
+add_custom_target(  displayEmulationCompileCommands ALL
```
```suggestion
add_custom_target(  displayEmulationCompileCommands
```

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,292 @@
+# Display the compile instructions in the simulation flow
+getCompileCommands("${COMMON_COMPILE_FLAGS}" "${SIMULATOR_COMPILE_FLAGS}" "${COMMON_LINK_FLAGS}" "${SIMULATOR_LINK_FLAGS}" "${SIMULATOR_TARGET}" "${SIMULATOR_OUTPUT_NAME}${EXECUTABLE_EXTENSION}")
+
+add_custom_target(  displaySimulationCompileCommands ALL
```
```suggestion
add_custom_target(  displaySimulationCompileCommands
```

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,292 @@
+# Display the compile instructions in the report flow
+getCompileCommands("${COMMON_COMPILE_FLAGS}" "${REPORT_COMPILE_FLAGS}" "${MODIFIED_COMMON_LINK_FLAGS_REPORT}" "${REPORT_LINK_FLAGS}" "${REPORT_TARGET}" "${REPORT_OUTPUT_NAME}${EXECUTABLE_EXTENSION}")
+
+add_custom_target(  displayReportCompileCommands ALL
```
```suggestion
add_custom_target(  displayReportCompileCommands 
```

yuguen commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/CMakeLists.txt:
```diff
@@ -0,0 +1,292 @@
+# Display the compile instructions in the fpga flow
+getCompileCommands("${COMMON_COMPILE_FLAGS}" "${FPGA_COMPILE_FLAGS}" "${COMMON_LINK_FLAGS}" "${FPGA_LINK_FLAGS}" "${FPGA_TARGET}" "${FPGA_OUTPUT_NAME}${EXECUTABLE_EXTENSION}")
+
+add_custom_target(  displayFPGACompileCommands ALL
```
```suggestion
add_custom_target(  displayFPGACompileCommands
```

yuguen reviewed (changes requested):
Forgot to fix these `ALL` in the cmake file

williamwang256 committed:
Apply suggestions from code review

Co-authored-by: yuguen-intel <yohann.uguen@intel.com>

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+# `Streaming Data Interfaces` Sample
+
+This FPGA sample is a tutorial that demonstrates how to implement streaming data interfaces on an IP component. If you have not already gone over the IP Authoring Interfaces Overview Tutorial, it is recommended that you do so before continuing with this tutorial.
```
```suggestion
This FPGA sample is a tutorial that demonstrates how to implement streaming data interfaces on an IP component. It is recommended that you review the IP Authoring Interfaces Overview Tutorial before continuing with this tutorial.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+## Purpose
+
+Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointer to elements within the FIFO.
```
```suggestion
Pipes are a first-in first-out (FIFO) buffer construct that provide data links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointers to elements within the FIFO.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointer to elements within the FIFO.
+
+The concept of a pipe provides us with a mechanism for specifying streaming data interfaces on an IP component. This tutorial will demonstrate how to declare and configure a pipe to implement your desired interface.
```
```suggestion
The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to declare and configure a pipe to configure an Avalon streaming interface.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+| Optimized for        | Description
+|:---                  |:---
+| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
+| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
```
```suggestion
| Hardware             | Intel® Agilex® 7, Cyclone 10® GX, Cyclone V®, Arria® 10, and Stratix® 10 FPGAs
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+### Configuring a Pipe to Implement a Streaming Data Interface
+
+Each individual pipe is a function scope class declaration of the templated `pipe` class.
+- The first template parameter is a user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the  RTL.
```
This would probably be clearer in a table:

| Template Parameter | Description
|---                 |---
| `name`             | User-defined type that differentiates this particular pipe from the others and provides a name for the interface in the RTL
| `dataT`            | Defines the datatype of elements carried by the interface on the `data` signal *
| `min_capacity`     | Specify a non-negative integer representing the capacity of the buffer on the input. This can help avoid some amount of bubbles in the pipeline in case the component itself stalls. This parameter is optional, and defaults to `0`.
| `properties`       | List of SYCL* `properties` that define additional semantic properties for a pipe. This parameter is optional.**

⚠️ * There is currently a known issue with using struct types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (e.g., unsigned char). For the time being, please use a wider datatype where applicable (e.g., 16-bits).

**A list of those properties relevant to this sample is given in Table 1 (please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications).

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+| `bits_per_symbol<int>`                  | non-negative integer that divides the size of the data type          | 8                                      |
+| `uses_valid<bool>`                      | boolean                                                              | `true`                                 |
+| `first_symbol_in_high_order_bits<bool>` | boolean                                                              | `true`                                 |
+| `protocol`                              | `protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready` | `protocol_avalon_streaming_uses_ready` |
```
```suggestion
| Property                                | Default Value                          |Valid Values
| ---                                     | ---                                    |---
| `ready_latency<int>`                    | 0                                      |non-negative integer
| `bits_per_symbol<int>`                  | 8                                      |non-negative integer that divides the size of the data type
| `uses_valid<bool>`                      | `true`                                 |boolean
| `first_symbol_in_high_order_bits<bool>` | `true`                                 |boolean
| `protocol`                              | `protocol_avalon_streaming_uses_ready` |`protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready`
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+| `first_symbol_in_high_order_bits<bool>` | boolean                                                              | `true`                                 |
+| `protocol`                              | `protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready` | `protocol_avalon_streaming_uses_ready` |
+
+> **Note:** These properties may be specified in *any* order within the oneAPI `properties` object. Omitting a single property from the properties class instructs the compiler to assume the default value for that property (i.e., you can just define the properties you would like to change from the default). Omitting the properties template parameter entirely instructs the compiler to assume the de ...
```
```suggestion
> **Note:** These properties may be specified in any order within the oneAPI `properties` list. Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for *all* properties.
```

**technically** the `properties` thing is a *type*, but I don't think that's terribly important to drive home in this tutorial, so I recommend just say "property list"

Eric has given me feedback before that it's bad practice to use abbreviations like "i.e." and "e.g." in documentation.

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+#### Example 1.
+
+The following example declares a pipe to implement a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.
```
```suggestion
The following example explicitly declares a pipe that implements a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
````diff
@@ -0,0 +1,272 @@
+The following example declares a pipe to implement a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.
+
+```c++
+// Unique user-defined types
````
```suggestion
// Forward declare the pipe name in the global scope. This is an FPGA best
// practice that reduces name mangling in the optimization reports.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+using FirstPipeInstance = sycl::ext::intel::experimental::pipe<
+    FirstPipeT,      // An identifier for the pipe
+    int,             // The type of data in the pipe
+    8,               // The capacity of the pipe
```
```suggestion
    8,               // Minimum capacity of the pipe (buffer depth)
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+You can enable Avalon streaming sideband signal support by using the special `StreamingBeat` struct provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the data type to your pipe. Only the `StreamingBeat` struct generates sideband signals when used with a pipe.
+
+The `StreamingBeat` struct is templated on three parameters.
```
a table would be nice here, similar to above

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+...
+
+using StreamingBeatDataT = sycl::ext::intel::experimental::StreamingBeat<unsigned char, true, true>;
```
```suggestion
using StreamingBeatDataT = sycl::ext::intel::experimental::StreamingBeat<unsigned short, true, true>;
```

You said earlier not to write code like this :)

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+#### Example 2.
+
+The following example shows how to configure a `StreamingBeat` struct to use `sop`, `eop` and `empty` by setting the second and third template parameters to `true`. Using the `StreamingBeatT` type defined here in a pipe will result in additional `startofpacket`, `endofpacket`, and `empty` signals on the interface.
```
```suggestion
The following example shows how to configure a `StreamingBeat` structure to use `sop`, `eop` and `empty` by setting the second and third template parameters to `true`. Using the `StreamingBeatT` type defined here in a pipe will result in `startofpacket`, `endofpacket`, and `empty` signals appearing on the resulting Avalon streaming interface.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,272 @@
+
+### Reading the Reports
+
+After compiling in the report flow, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pa ...
```
```suggestion
After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pane. This pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).
```

whitepau reviewed (changes requested):
great looking code sample!

williamwang256 committed:
Apply suggestions from Pau;scode review

Co-authored-by: Paul White <paul.white@intel.com>

williamwang256 committed:
addressing comments, readme work

williamwang256 committed:
add sample to top-level README

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+Pipes are a first-in first-out (FIFO) buffer construct that provide data links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointers to elements within the FIFO.
+
+The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to declare and configure a pipe to configure an Avalon streaming interface.
```
```suggestion
The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to use the pipe API to configure an Avalon streaming interface.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+### Configuring a Pipe to Implement a Streaming Data Interface
+
+Each individual pipe is a function scope class declaration of the templated `pipe` class. It takes two mandatory and two optional parameters, as summarized in Table 1.
```
```suggestion
Each pipe is a class declaration of the templated `pipe` class. A pipe declaration takes two mandatory and two optional parameters, as summarized in Table 1.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+| Template Parameter                      | Description
+| ---                                     | ---
+| `name`                                  | A user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the RTL.
```
```suggestion
| `name`                                  | A user-defined type that uniquely identifies this pipe. The name of this type is also used to identify the interface in the generated RTL.
```

williamwang256 committed:
Minor updates to README wording

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+| ---                                     | ---
+| `name`                                  | A user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the RTL.
+| `dataT`                                 | The datatype of elements carried by the interface.*
+| `min_capacity`                          | A non-negative integer representing the capacity of the buffer on the input. This can help avoid some amount of bubbles in the pipeline in case the component itself stalls. This parameter is optional, and defaults to 0.
```
```suggestion
| `dataT`                                 | The type of data that passes through the pipe*. This is the data type that is read during a successful pipe read() operation, or written during a successful pipe write() operation. The type must have a standard layout and be trivially copyable.
| `min_capacity`                          | User-defined minimum number of words (in units of `dataT`) that the pipe must be able to store without any being read out. This parameter is optional, and defaults to 0.
```

https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/2023-2/the-pipe-class-and-its-use.html

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+> :warning: * There is currently a known issue with using structure types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (for example, `unsigned char`). For the time being, please use a wider datatype where applicable (for example, `unsigned short`).
+
+Below is a summary of all relevant SYCL properties which can be applied to a `pipe` using the fourth template parameter. Please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other app ...
```
```suggestion
Below is a summary of all relevant SYCL properties which can be applied to a `pipe` using the `properties` template parameter. Please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications.
```

it might be nice to link to the avalon interface specifications https://www.intel.com/content/www/us/en/docs/programmable/683091/22-3/st-interface-properties.html but maybe that will make this too long and that should really be in the manual.

williamwang256 replied:
I just added a link to the Avalon interface spec after the table:

_See [this page](https://www.intel.com/content/www/us/en/docs/programmable/683091/current/st-interface-properties.html) for more information on the Avalon interface specifications._

Let me know if you think we should keep it.

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+| `protocol`                              | `protocol_avalon_streaming_uses_ready` |`protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready`
+
+
+** Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for all properties.
```
```suggestion
> *Note*: ** Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for all properties.
```
This can be moved right under the note about the unsigned char bug

williamwang256 committed:
Fix typo

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+// Pipe properties (listed here are the defaults; this achieves the same
+// behavior as not specifying any of these properties)
+using PipePropertiesT = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::ready_latency<0>,
+    sycl::ext::intel::experimental::bits_per_symbol<8>,
+    sycl::ext::intel::experimental::uses_valid<true>,
+    sycl::ext::intel::experimental::first_symbol_in_high_order_bits<true>,
+    sycl::ext::intel::experimental::protocol_avalon_streaming_uses_ready));
```
Is it `decltype (properties { ready_latency...})`? I've seen the `properties` type configured with curly braces and regular parentheses.

williamwang256 replied:
interesting, both seem to work... however our documentation shows it with regular parentheses, so I will probably stick to that. (e.g. https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/2023-2/host-pipe-declaration.html)

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+    int,             // The type of data in the pipe
+    8,               // Minimum capacity of the pipe (buffer depth)
+    PipePropertiesT  // Customizable pipe properties
+    >;
```
```suggestion
using FirstPipe = sycl::ext::intel::experimental::pipe<
    FirstPipeT,      // An identifier for the pipe
    int,             // The type of data in the pipe
    8,               // Minimum capacity of the pipe (buffer depth)
    PipePropertiesT  // Customizable pipe properties
    >;
```

it's not technically an instance

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+### Avalon Streaming Sideband Signals
+
+You can enable Avalon streaming sideband signal support by using the special `StreamingBeat` structure provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the data type to your pipe. Only the `StreamingBeat` structure generates sideband signals when used with a pipe.
```
```suggestion
Pipes support a subset of Avalon streaming sideband signals. You can add these to your pipe interface by using the special `StreamingBeat` structure provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the `dataT` of your pipe. Only the `StreamingBeat` structure generates sideband signals when used with a pipe.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+| Template Parameter                      | Description
+| ---                                     | ---
+| `dataT`                                 | The datatype of elements carried by the interface.
```
```suggestion
| `dataT`                                 | The datatype of elements carried by the `data` signal of the Avalon streaming interface.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+| Template Parameter                      | Description
+| ---                                     | ---
+| `dataT`                                 | The datatype of elements carried by the interface.
+| `uses_packets`                          | A boolean to indicate whether to enable additional `startofpacket` (`sop`) and `endofpacket` (`eop`) signals to the Avalon interface.
```
```suggestion
| `uses_packets`                          | A boolean that indicates whether to enable the `startofpacket` (`sop`) and `endofpacket` (`eop`) sideband signals on the Avalon streaming interface.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+| ---                                     | ---
+| `dataT`                                 | The datatype of elements carried by the interface.
+| `uses_packets`                          | A boolean to indicate whether to enable additional `startofpacket` (`sop`) and `endofpacket` (`eop`) signals to the Avalon interface.
+| `uses_empty`                            | A boolean to indicate whether to enables the `empty` signal, which indicates the number of symbols that are empty during the `eop` cycle.
```
```suggestion
| `uses_empty`                            | A boolean that indicates whether to enable the `empty` sideband signal on the Avalon streaming interface
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+## Understanding the Tutorial
+
+In `threshold_packets.cpp`, two pipes are declared for implementing the streaming input and streaming output interfaces on a kernel which thresholds pixel values in an image. The streams use start of packet and end of packet signals to determine the beginning and end of the image.
```
```suggestion
In `threshold_packets.cpp`, two pipes are declared to implement the input and output streaming interfaces on a kernel which thresholds pixel values in an image. The streams use `startofpacket` and `endofpacket` signals to determine the beginning and end of the image.
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md:
```diff
@@ -0,0 +1,289 @@
+
+### Reading the Reports
+
+After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details p ...
```
```suggestion
After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information about these interfaces in the 'details' pane. The 'details' pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to size of the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+
+#include "exception_handler.hpp"
+
+#define THRESHOLD 200
```
```suggestion
// limit pixel values to this value, or less
#define THRESHOLD 200
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+class OutPixel;
+class Threshold;
+
+// StreamingBeat struct to support Avalon packet sideband signals
```
```suggestion
// StreamingBeat struct enables sideband signals in Avalon streaming interface
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+class Threshold;
+
+// StreamingBeat struct to support Avalon packet sideband signals
+using StreamingBeatT = sycl::ext::intel::experimental::StreamingBeat<
+    unsigned short, true, false>;
```
```suggestion
using StreamingBeatT = sycl::ext::intel::experimental::StreamingBeat<
    unsigned short, // type carried over this Avalon streaming interface's data signal
    true,         // enable startofpacket and endofpacket signals
    false>;      // disable the empty signal
```
elaborate magic numbers

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+      // Write out result
+      StreamingBeatT out_beat(pixel, start_of_packet, end_of_packet);
+      OutPixelPipe::write(out_beat);
+
```
```suggestion
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+template <typename InPixelPipe, typename OutPixelPipe>
+struct ThresholdKernel {
+
+  void operator()() const {
```
video IPs like this will generally either use a Streaming invocation interface, or a start/stop CSR interface like i used here:
https://github.com/intel-sandbox/whitepau.csc-oneapi/blob/mm_agents/src/csc_kernel.hpp

I'm on the fence about whether you should change it though; that extra code distracts from the focus of this tutorial, so I would lean to saying you should leave it as is.

PS i'm sorry about all the empty lines in that code; i don't know why clang-format did that.

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+    bool end_of_packet = false;
+
+    while (!end_of_packet) {
+
```
```suggestion
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+      PipePropertiesT // Customizable pipe properties
+      >;
+  using OutPixelPipe = sycl::ext::intel::experimental::pipe<
+      OutPixel,       // An identifier for the pipe
```
```suggestion
      OutStream,       // An identifier for the pipe
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+
+  // Image streams
+  using InPixelPipe = sycl::ext::intel::experimental::pipe<
+      InPixel,        // An identifier for the pipe
```
```suggestion
      InStream,        // An identifier for the pipe
```

whitepau commented on DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/src/streaming_data_interfaces.cpp:
```diff
@@ -0,0 +1,119 @@
+    sycl::ext::intel::experimental::protocol_avalon_streaming_uses_ready));
+
+  // Image streams
+  using InPixelPipe = sycl::ext::intel::experimental::pipe<
+      InPixel,        // An identifier for the pipe
+      StreamingBeatT, // The type of data in the pipe
+      0,              // The capacity of the pipe
+      PipePropertiesT // Customizable pipe properties
+      >;
+  using OutPixelPipe = sycl::ext::intel::experimental::pipe<
+      OutPixel,       // An identifier for the pipe
+      StreamingBeatT, // The type of data in the pipe
+      0,              // The capacity of the pipe
+      PipePropertiesT // Customizable pipe properties
+      >;
```
Please declare these pipes at the global scope up with your kernel. This also lets you directly use the pipes in your kernel without needing the template parameters.

whitepau reviewed (changes requested):
nice code sample. minor changes.

whitepau reviewed (commented):
I made a couple mistakes

williamwang256 committed:
Apply suggestions from code review

Co-authored-by: Paul White <paul.white@intel.com>

williamwang256 committed:
tidying things up and addressing more comments from review

williamwang256 committed:
minor updates

williamwang256 committed:
update links

williamwang256 committed:
minor wording changes

yuguen committed:
Merge master branch into development (#1861)

aaronkintel committed:
FPGA: Change template unrolls to pragma in ANR (#1862)

## Description

Changed preprocessor unrolls to `#pragma unroll` when the compiler uses `ivdep(array)`.

# Fixes Issue

Compiler could not achieve II=1 for some loops since `ivdep(array)` could not penetrate lambdas from the unroller.

williamwang256 committed:
more fixes from review

williamwang256 committed:
add link to avalon spec

williamwang256 committed:
FPGA: Update Buffer Locations in Matrix Multiply sample to start from 1 (#1864)

Updating buffer locations in Matrix Multiply sample to match requirements (BLs start at 1 instead of 0) for 2024.0 release.

justin-rosner committed:
FPGA: Remove use of get_pointer from lsu_control sample (#1860)

This PR updates all uses of get_pointer to get_multi_ptr as there is now a deprecation warning when using the latest sycl/rel version.

williamwang256 committed:
move to ip_authoring_interfaces folder

williamwang256 committed:
Merge branch 'streaming-interfaces' of https://github.com/williamwang256/oneAPI-samples into streaming-interfaces

williamwang256 committed:
minor README updates

whitepau committed:
FPGA: Move the platform_designer sample out of experimental (#1858)

Move Platform Designer out of experimental directory

williamwang256 committed:
Update DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/streaming_data_interfaces/CMakeLists.txt

Co-authored-by: yuguen-intel <yohann.uguen@intel.com>

yug-intel committed:
FPGA: Passing Functor Directly to Lambda (#1865)

Passing the functor object directly to the lambda.

yuguen committed:
FPGA: Remove mentions to now deprecated PAC boards (#1868)

Remove all mentions to the PAC boards which are going to be deprecated in the 2024.0 release.

juny-lim committed:
FPGA: Add customized DSP RTL code to use_library sample (#1870)

Added customized DSP RTL code to src folder.
Updated CMakeList to use template from fpga_template.
Updated ReadMe
Showcase specified RTL resource in XML

rho180 committed:
FPGA: update io_streaming sample README (#1875)

The code sample includes a macro (USE_REAL_IO_PIPES) that was previously undocumented in the README. This PR corrects this issue.

fredderf121 committed:
FPGA: Rewrite of Max Interleaving Code Sample So That Interleaving Is Possible (#1879)

The loops in the old code sample had memory dependencies that prevented interleaving from occurring and relied on the compiler's inability to analyze memory dependencies to demonstrate the area savings of interleaving.
An upcoming internal compiler PR improves our compiler's analysis abilities, rendering the old sample obsolete.

As such, the code sample has been modified to contain loops that support interleaving. The README.md has been updated with the new hardware runtime data and some structural changes, as the old README.md frequently stated how the loops did not support interleaving, which is no longer the case.

yuguen committed:
Update DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/streaming_data_interfaces/src/streaming_data_interfaces.cpp

yuguen committed:
Update DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/streaming_data_interfaces/src/streaming_data_interfaces.cpp

yuguen committed:
Update DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/streaming_data_interfaces/src/streaming_data_interfaces.cpp

yuguen committed:
FPGA: add missing fstream include to PCA and MVDR Beamforming (#1883)

This change adds an include which is now necessary as not included by default with sycl.hpp anymore.

krisrak committed:
updated Jupyter GPU Opt (#1884)

ajaykumarkannan committed:
FPGA: Add sycl:: prefix to math functions in crr, mvdr, and read_only_cache (#1882)

This PR adds sycl:: prefix to math functions to ensure that we pick up the optimized versions of the math functions as compared to the std:: versions.

KevinUTAT committed:
FPGA: add the -xsparallel=2 flag to the fpga_compile sample (#1885)

Adding -Xsparallel=2 to CMakeLists to fix compilation issue

KevinUTAT committed:
FPGA: Add new sample `component_interfaces_comparison` (#1878)

This sample introduces different invocation/data interfaces that can be used in IP Authoring flow

yuguen committed:
FPGA: New Reference design code sample: `fft2d` (#1871)

This change adds a new FPGA code sample implementing a 2D FFT on Intel FPGAs.

yuguen committed:
Merge branch 'development' into streaming-interfaces
